pkg/devfile/adapters/common/command.go

```diff
@@ -13,66 +13,70 @@ import (
 func getCommand(data data.DevfileData, commandName string, groupType common.DevfileCommandGroupType) (supportedCommand common.DevfileCommand, err error) {
 
 	commands := data.GetCommands()
+	var firstMatchedCommand common.DevfileCommand
 
-	for _, command := range commands {
-
-		// validate command
-		err = validateCommand(data, command)
-
+	if commandName == "" {
+		// validate the command groups before searching for a command match
+		// if the command groups are invalid, err out
+		// we only validate when the push command flags are absent,
+		// since we know the command kind from the push flags
+		err = validateCommandsForGroup(data, groupType)
 		if err != nil {
 			return common.DevfileCommand{}, err
 		}
+	}
 
+	for _, command := range commands {
 		// if command is specified via flags, it has the highest priority
 		// search through all commands to find the specified command name
 		// if not found fallback to error.
 		if commandName != "" {
 
-			// Update Group only custom commands (specified by odo flags)
-			command = updateGroupforCommand(groupType, command)
-
 			if command.Exec.Id == commandName {
 
+				// Update Group only custom commands (specified by odo flags)
+				command = updateCommandGroupIfReqd(groupType, command)
+
 				// we have found the command with name, its groupType Should match to the flag
 				// e.g --build-command "mybuild"
 				// exec:
 				//   id: mybuild
 				//   group:
 				//     kind: build
 				if command.Exec.Group.Kind != groupType {
-					return supportedCommand, fmt.Errorf("command group mismatched, command %s is of group %v in devfile.yaml", commandName, command.Exec.Group.Kind)
+					return command, fmt.Errorf("command group mismatched, command %s is of group %v in devfile.yaml", commandName, command.Exec.Group.Kind)
 				}
-				supportedCommand = command
-				return supportedCommand, nil
+
+				return command, validateCommand(data, command)
 			}
 			continue
 		}
 
 		// if no command specified via flag, default command has the highest priority
-		// We need to scan all the commands to find default command
 		// exec.Group is a pointer, to avoid null pointer
-		if command.Exec.Group != nil && command.Exec.Group.Kind == groupType && command.Exec.Group.IsDefault {
-			supportedCommand = command
-			return supportedCommand, nil
+		if command.Exec.Group != nil && command.Exec.Group.Kind == groupType {
+			if command.Exec.Group.IsDefault {
+				// We need to scan all the commands to find default command
+				return command, validateCommand(data, command)
+			} else if reflect.DeepEqual(firstMatchedCommand, common.DevfileCommand{}) {
+				// store the first matched command in case there is no default command
+				firstMatchedCommand = command
```

Comment:
should we rename it to `onlyCommand` or `singleCommand`? IIUC we are throwing error if more than one command of a grouptype is present and none of them is default and if only one command is present without any default value we are allowing it.  So it should be that one command

Reply:
yeah we can rename it.

Actually there is a case where multiple defaults can be present ie when we use `odo push --build-command="mybuild1"`. `mybuild1` and `mybuild2` can be default or have no defaults at all and this is fine since we're explicitly using a flag and saying what command to run

Reply:
but that case we are already covering above if commandName != "" , IIUC  that case would not fallback here, also can we update the comment 	`// store the first matched command in case there is no default command` to `return the only remaining command for the group`

Reply:
Not necessary to do in this PR, but should we separate the flag command and devfile command function to better code readability.
`getCommandFromFlags`
`getCommandfromDevfile`
and in the `GetRunCommand`, `GetBuildCommand` functions includes the check for command name?
IMO this one(what i had written also) is difficult to understand.

Reply:
> but that case we are already covering above if commandName != "" , IIUC that case would not fallback here, also can we update the comment // store the first matched command in case there is no default command to return the only remaining command for the group

you're right, i have updated the comment to 
```
// return the only remaining command for the group if there is no default command
// NOTE: we return outside the for loop since the next iteration can have a default command
```

> Not necessary to do in this PR, but should we separate the flag command and devfile command function for better code readability?

yeah i think it makes sense to have different funcs and separate out their logic for better readability. I think anyone else reading it will be confused in the first few attempts.

Reply:
I've updated the getCommand() func. Broken down into getCommandFromFlag() and getCommandfromDevfile() and moved their specific logic to this func.

getCommand() does the check for commandName and splits into either of the above two function calls. This way we can just use getCommand() from anywhere instead of doing this check everywhere when we need a command.

Reply:
@maysunfaisal Thanks for doing this, now it is much easier to understand :)